### PR TITLE
Redesign amit

### PR DESCRIPTION
**Pull request by amrx101**
```
Running `target/debug/mqttwrk bench -c 2 -n 1000000`
Connections & subscriptions ok. Elapsed = 0
[00:02:49] #######################----------------- 1114271/2000000 
Id = rumqtt-sink-00000
            Incoming publishes : Received = 1000000 Throughput = 5906.43 messages/s
[00:09:24] ######################################## 1996600/2000000 
Id = rumqtt-00000
            Outgoing publishes : Received = 1000000 Throughput = 1770.971 messages/s
            Incoming publishes : Received = 0       Throughput = 0 messages/s
            Reconnects         : 0
# of samples          : 884408
99.999'th percentile  : 180
99.99'th percentile   : 165
90 percentile         : 7
[00:09:25] ######################################## 1998434/2000000 
Id = rumqtt-00001
            Outgoing publishes : Received = 1000000 Throughput = 1767.1091 messages/s
            Incoming publishes : Received = 0       Throughput = 0 messages/s
            Reconnects         : 0
# of samples          : 885467
99.999'th percentile  : 182
99.99'th percentile   : 135
90 percentile         : 7
50 percentile         : 2
Aggregate
99.999'th percentile  : 180
99.99'th percentile   : 145
90 percentile         : 7
```

### CODE REVIEW

File: src/link.rs [hunk 1]
```diff
@@ -18,6 +22,9 @@ impl Link {
         ca_file: Option<String>,
         client_cert_file: Option<String>,
         client_key_file: Option<String>,
+        indi_sender: Option<Sender<i32>>,
```
Comment (tekjar) [inline src/link.rs]: Both the senders can be combined. 

```rust
status_tx: Sender<Status>
```

```
enum Status {
    Progress(usize),
    Histogram(Histogram)
}
```

Also I'm not sure if `Option` is necessary.

File: src/bench/connection.rs [hunk 2]
```diff
@@ -172,6 +170,14 @@ impl Connection {
                                 None => warn!("No publish record for Pkid={:?}", index),
                             };
 
+                            // send progress tick over indi_sender
+                            match &self.link.indi_sender.clone(){
```
Comment (tekjar) [inline src/bench/connection.rs]: This is not correct. Option which has sender is cloned and referenced only to clone it again in Some again. `Sender` implements `Sync + Send`. Why is it being cloned for every `send`?

File: src/bench/connection.rs [hunk 3]
```diff
@@ -172,6 +170,14 @@ impl Connection {
...
+                                Some(v) => {
+                                    v.clone().send(1).await.unwrap();
```
Comment (tekjar) [inline src/bench/connection.rs]: Sending progress for every ack doesn't scale for a lot of connections. We should maintain a local counter and periodically send progress.

```rust
loop {
    select! {
        v = eventloop.poll() => { // collect progress }
        _ = periodic_timer => { // send progress }
    }
}
```

**Review by tekjar — CHANGES_REQUESTED** [review]

File: src/link.rs (hunk 1 above)
Comment (amrx101) [reply src/link.rs]: Done

File: src/bench/connection.rs [hunk 4]
```diff
@@ -127,37 +136,72 @@ impl Connection {
...
-            let event = match self.link.eventloop.poll().await {
-                Ok(v) => v,
-                Err(e) => {
-                    error!("Id = {}, Connection error = {:?}", self.link.id, e);
-                    reconnects += 1;
-                    if reconnects == 1 {
-                        break;
+            tokio::select!{
+                event = self.link.eventloop.poll() => {
+                    let mm = match event {
```
Comment (tekjar) [inline src/bench/connection.rs]: change `let mm` to `let event` or some other better name

File: src/bench/connection.rs [hunk 5]
```diff
@@ -127,37 +136,72 @@ impl Connection {
...
-                    Incoming::PubAck(_pkid) => acks_count += 1,
-                    Incoming::Publish(_publish) => incoming_count += 1,
-                    Incoming::PingResp => {}
-                    incoming => {
-                        error!("Id = {}, Unexpected incoming packet = {:?}", id, incoming);
-                        break;
+                    match mm {
+                        Event::Incoming(v) => {
+                            match v {
+                                Incoming::PubAck(_pkid) => {
```
Comment (tekjar) [inline src/bench/connection.rs]: Remove `_` from pkid.

File: src/bench/mod.rs [hunk 6]
```diff
@@ -3,20 +3,30 @@ use futures::StreamExt;
...
     let config = Arc::new(config);
     let barriers_count = config.connections + config.sink;
     let barrier = Arc::new(Barrier::new(barriers_count));
     let mut handles = futures::stream::FuturesUnordered::new();
+    let total_expected = config.count * config.publishers * config.connections;
+    let sty = ProgressStyle::default_bar()
+        .template("[{elapsed_precise}] {bar:40.cyan/blue} {pos:>7}/{len:7} {msg}")
+        .progress_chars("##-");
 
+    let (c_tx, c_rx) = async_channel::unbounded::<Status>();
```
Comment (tekjar) [inline src/bench/mod.rs]: Change `c_tx` and `c_rx` to `status_tx`, `status_rx`

File: src/bench/mod.rs [hunk 7]
```diff
@@ -51,6 +62,7 @@ pub(crate) async fn start(config: BenchConfig) {
         }));
     }
 
+    let (cc_tx, _) = async_channel::unbounded::<Status>();
```
Comment (tekjar) [inline src/bench/mod.rs]: If there is no receiver, why is there a channel?

Comment (tekjar) [reply src/bench/mod.rs]: Need a better name than `cc_tx`

File: src/bench/mod.rs [hunk 8]
```diff
@@ -75,9 +88,45 @@ pub(crate) async fn start(config: BenchConfig) {
         }));
     }
 
+    let pb = ProgressBar::new(total_expected as u64);
+    pb.set_style(sty.clone());
+    let mut r_cnt = 0;
```
Comment (tekjar) [inline src/bench/mod.rs]: Need a better name than `r_cnt`.

**Review by tekjar — CHANGES_REQUESTED** [review]

**Comment by tekjar** [comment]
Left some nits. Also is this tested? Every PR should be tested with atleast a 100 connections.

**Comment by amrx101** [comment]
> Left some nits. Also is this tested? Every PR should be tested with atleast a 100 connections.

tested with 100 conns with 10000 messages each con

**Review by tekjar — APPROVED** [review]

**Pull request merged**